packages/flutter/lib/src/painting/shape_decoration.dart

```diff
@@ -336,19 +336,28 @@ class _ShapeDecorationPainter extends BoxPainter {
       _interiorPaint!.shader = _decoration.gradient!.createShader(rect, textDirection: textDirection);
     }
     if (_decoration.shadows != null) {
-      if (_shadowCount == null) {
-        _shadowCount = _decoration.shadows!.length;
-        _shadowPaints = <Paint>[
-          ..._decoration.shadows!.map((BoxShadow shadow) => shadow.toPaint()),
+      _shadowCount ??= _decoration.shadows!.length;
+      if (_decoration.shape.preferPaintInterior) {
+        _shadowBounds = <MapEntry<Rect, Paint>>[
+          ..._decoration.shadows!.map((BoxShadow shadow) {
+            return MapEntry<Rect, Paint>(
+              rect.shift(shadow.offset).inflate(shadow.spreadRadius),
+              shadow.toPaint(),
+            );
+          }),
+        ];
+      } else {
+        _shadowPaths = <MapEntry<Path, Paint>>[
+          ..._decoration.shadows!.map((BoxShadow shadow) {
+            return MapEntry<Path, Paint>(
+              _decoration.shape.getOuterPath(rect.shift(shadow.offset).inflate(shadow.spreadRadius), textDirection: textDirection),
+              shadow.toPaint(),
+            );
+          }),
```

Comment:
i guess technically we're computing the paints more than we need to (we only need to compute them the first time, not when rect changes). Maybe add a TODO? Or not, your call. It's not a big deal.

Reply:
I need test, but what if rotation happens? Is it the same shadow?

Reply:
I'll try later (possibly tomorrow) but if get this working, optimization can be really big.

Reply:
well the _paint_ definitely isn't affected by the rect (it's computed by `shadow.toPaint()`, which doesn't get the rect). But the question is whether it's more expensive to compute the paint, or more expensive to handle the overhead of storing the paint in some other data structure in parallel with the rect bounds / paths. IIRC you had it the other way before and it was slower?

Reply:
I believe it was slower because of the additional loop, not exactly the additional structure. In this case, everything is going to be mostly similar. But I'll test, let's see.

Reply:
One thing we could also do is not-cache the shadow path for primitive types (rectangle/stadium/circle), so later we can call `paintInterior` for the shadow. Your question will be answered with this. If performance is better, it is `drawPath`, if performance is worse, it is `cachePath` that takes longer.

Reply:
I have a proposal.

What if we had a `mixin PrimitiveShape { paintInterior(...) } ` that is applied on RoundedRectangleBorder / CircleBorder / Stadium Border, then we do `if (shape is PrimitiveShape) { (shape as PrimitiveShape).paintInterior(...) }`?

We wouldn't need `preferPaintInterior`. And... maybe, one day, we could delimit the `BorderSide.only`/`MultipleBorderSide` (I still need to convince you) to only work on shapes that have that mixin. Is this too crazy?

Reply:
Update: your proposal already happens. `interiorPaint` is only set when not null. Gradient, image and shadow depend on the rect. So it is already super optimized. 

<img width="830" alt="image" src="https://user-images.githubusercontent.com/351125/182449700-e087bb52-cad7-41e2-9f34-74af4b419823.png">

Reply:
Found what you said. Now I understand the issue. Testing a few things.

Reply:
After debugging today: negligible difference, I'm reverting for the old way. Not changing what already works fine. My sample only uses a single color, if paint is complex enough, there might be differences.

Again, you are correct. Thanks!